perf_benchmark/benchmark_common.cc

```diff
@@ -0,0 +1,294 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+////////////////////////////////////////////////////////////////////////////////
+//
+#include <fstream>
+#include <sstream>
+#include <utility>
+#include "perf_benchmark/benchmark_common.h"
+#include "google/protobuf/text_format.h"
+#include "absl/strings/escaping.h"
+#include "nlohmann/json.hpp"
+
+namespace google {
+namespace grpc {
+namespace transcoding {
+
+namespace perf_benchmark {
+
+namespace pb = ::google::protobuf;
+
+std::string LoadFile(const std::string& file_name) {
+  std::ifstream ifs(file_name);
+  if (!ifs) {
+    std::cerr << "Could not open " << file_name.c_str() << std::endl;
+    return {};
+  }
+  std::ostringstream ss;
+  ss << ifs.rdbuf();
+  return ss.str();
+}
+
+bool LoadService(const std::string& config_pb_txt_file,
+                 ::google::api::Service* service) {
+  static const char kBenchmarkData[] = "perf_benchmark/";
+  return LoadService(config_pb_txt_file, kBenchmarkData, service);
+}
+
+bool LoadService(const std::string& config_pb_txt_file,
+                 const std::string& benchmark_path,
+                 ::google::api::Service* service) {
+  auto config = LoadFile(benchmark_path + config_pb_txt_file);
+  if (config.empty()) {
+    return false;
+  }
+
+  if (!pb::TextFormat::ParseFromString(config, service)) {
+    std::cerr << "Could not parse service config from "
+              << config_pb_txt_file.c_str() << std::endl;
+    return false;
+  } else {
+    return true;
+  }
+}
+double GetPercentile(const std::vector<double>& v, double perc) {
+  if (perc < 0) { perc = 0; }
+  if (perc > 100) { perc = 100; }
+  // Making a copy since std::nth_element mutates the vector
+  auto copy = std::vector<double>(v);
+  size_t rough_position = copy.size() * perc / 100;
+  std::nth_element(copy.begin(), copy.begin() + rough_position, copy.end());
+  return copy[rough_position];
+}
+
+std::string GetRandomBytesString(int64_t length, bool base64) {
+  static const char charset[] =
+      {0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20,
+       21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38,
+       39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56,
+       57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67, 68, 69, 70, 71, 72, 73, 74,
+       75, 76, 77, 78, 79, 80, 81, 82, 83, 84, 85, 86, 87, 88, 89, 90, 91, 92,
+       93, 94, 95, 96, 97, 98, 99, 100, 101, 102, 103, 104, 105, 106, 107, 108,
+       109, 110, 111, 112, 113, 114, 115, 116, 117, 118, 119, 120, 121, 122,
+       123, 124, 125, 126, 127};
+  std::string ret;
+  ret.reserve(length);
+
+  for (int i = 0; i < length; ++i) {
+    ret += charset[rand() % (sizeof(charset) - 1)];
+  }
+  return base64 ? absl::Base64Escape(ret) : ret;
+}
+
+std::string GetRandomString(int64_t length) {
+  static const char charset[] =
```

Comment:
nit: why are only these characters used? Is there any concern with allowing any valid character?

Reply:
Great question! Because of JSON format, we'd have to escape `"` and `\`. Otherwise, the parsing would stop earlier since it will see `"` as the end of string.
I'm only using alphanumeric string just for simplicity. The `bytes` typed proto covers complete set of ASCII characters.